### DISCUSSION TITLE
Have you considered using `Zod` for creating static types?

### DISCUSSION THREAD

**Discussion by timbrinded** [wevm/viem#567] [Idea / Feature Request] [answered]
Are you guys aware of [Zod](https://github.com/colinhacks/zod) or libraries like it to make typescript actually typesafe at runtime not just compiletime?

The general idea is you create schemas containing rules and route your logic through zod helpers which not only give you errors where you expect, but automatically generates types for you. 

```
import { z } from "zod";

// creating a schema for strings
const mySchema = z.string();

// parsing
mySchema.parse("tuna"); // => "tuna"
mySchema.parse(12); // => throws ZodError
```

So for example, viem often uses the `0x${string}` literal type which is fine in small projects but completely breaks in larger projects that are unaware of a very viem-opinionated concept. Instead you would use strings but have a rule defined as must start with 0x.

Similarly the clients would be massively simplified by not having the heavily inferred types supported by reams of type helpers

**Comment by tmm** [comment, marked as answer]
We are familiar! viem's type inference is powered by our [ABIType](https://github.com/wagmi-dev/abitype) and it exports [Zod schemas](https://abitype.dev/api/zod.html) for the core ABI types.

In key areas, we already validate requests, transactions, etc. at runtime. It's unlikely that we would add extensive additional runtime validation without a specific reason since it will slow things down (Zod isn't very fast [compared to other runtime validators](https://github.com/moltar/typescript-runtime-type-benchmarks/)) and most folks use viem with TypeScript anyway.

–

> So for example, viem often uses the `` `0x${string}` `` literal type which is fine in small projects but completely breaks in larger projects

Say more. Is it cumbersome to assert `string`s to `` `0x${string}` `` or something else? You can [configure the internal types](https://abitype.dev/config.html) to use whatever you want.

> Similarly the clients would be massively simplified by not having the heavily inferred types supported by reams of type helpers

Agree that they would be simplified, but we would trade that for worse developer experience (e.g. no more type inference for ABI function/event names, arguments, EIP-721 typed data, etc.)

**Reply by timbrinded** [reply]
My bad, my original post is like teaching grandma to suck eggs!

I'll look into AbiType customization to see if it makes my life easier
